Testing

AaronJai opened the pull request:
Opening this PR in preparation for merge.

**Core Purpose**
- This branch focuses on implementing both unit tests and Selenium tests.
- Some tests have already been added — feel free to contribute more if you have ideas.

**Merge Context**
- As of now, ```sharepage_leaderboard``` has been merged in.
- We're still waiting on ```viewdata_rank``` before merging this branch into ```main```.

**Additional Suggestions**
- While working on this branch, I recommend everyone review and refactor their own work for clarity and consistency.
E.g.,:
- Consolidating repeated Tailwind styles into reusable class names
- moving scripts from html files into proper javascript files in the ```static``` folder
- Creating a ```controllers.py``` file for cleaner route logic (see #44)

useri4l committed:
feat: add leaderboard to share page

useri4l committed:
feat: dynamic preview display and layout fixes in leaderboard

useri4l committed:
feat: preview display for shared leaderboard

useri4l committed:
fix: hide View Details for current user in leaderboard

useri4l committed:
feat: current user in shared leaderboard

useri4l committed:
feat: current user in shared leaderboard

useri4l committed:
fix:float precision display

AaronJai committed:
fix: address #50 where incorrect data is shown
- issue was incorrect calculations. Used the same calculation as we did in dashboard
- flipped order of home and food summary card to reflect dashboard for consistency

useri4l committed:
feat: update currently share with section

AaronJai committed:
feat: allow toggling of view details on share page
- they can now close it after opening
- fixed a bug where initial visit required two clicks to open details; caused by redundant  div interfering with event handling

AaronJai committed:
fix #52: ensure progress bar line exceed further than 100%

AaronJai committed:
style: removed user info in preview card to remove double up and background highlighting to selected shared user

useri4l committed:
feat:Unread message notification

useri4l committed:
cleanup: deleted unnecessary comments

useri4l committed:
style: consistent indentation and spacing

useri4l committed:
Merge branch 'main' into sharepage_leaderboard

AaronJai committed:
chore: merge Alembic heads and regenerate Tailwind output.css

AaronJai committed:
fix #57: confirm email route was calling incorrect function
- function name might've changed in models.py when working on the profile page

AaronJai committed:
chore: remove debug console log

useri4l committed:
feat: add leaderboard section

useri4l committed:
feat: Add chart comparing between two users

AaronJai committed:
style: added cursor-pointer to some sections and retain x-axis font when comparing to other users

AaronJai committed:
refactor: restructure app to support testing
- created new configuration options
- create factory pattern
- create blueprints

AaronJai committed:
refactor: update html file url_for endpoints to include "main."
- had to also refactor how the sidebar handles active highlighting to accomodate endpoint change to include  "main."

MeiLin99999 committed:
test: add unit tests for user model password hashing

useri4l committed:
feat: add pie chart analysis display

Add travel, home, food, and shopping percentage analysis.
Implemented percentage breakdowns for each category using frontend calculations.

useri4l committed:
fix: restore accidentally deleted function

useri4l committed:
chore: clean up code indentation

AaronJai committed:
Merge branch 'testing' into sharepage_leaderboard

useri4l committed:
fix: make viewdata responsive and update CSS

useri4l committed:
refactor: migrate chart summary to backend

- Moved travel/home/food/shopping insight generation from js to route.py
- Simplified front-end template to render server-calculated data
- Fixed responsiveness for small/medium screens

AaronJai committed:
feat: create new unittests
- moved test folder to root of project
- new readme contains more info on how to do testing
- still need to merge Hazel's branch containing the new message models

AaronJai committed:
refactor: move dashboard category card calculation to backend
- for consistency and safety, similar to share page
- created new API endpoint to handle calculations

useri4l committed:
fix: adjust output.css

useri4l committed:
feat: dynamically update suggestions based on emission data

AaronJai committed:
fix: clicking category card on dashboard was redirecting but not scrolling down.
- selector needed to be updated; might've been changed in a previous commit

AaronJai committed:
feat: add basic selenium test class and update requirements file

AaronJai committed:
feat: update selenium to include sign up and sign in, and update readme files for testing

AaronJai committed:
reafactor: move app creation, imports, and model references to setUpClass

AaronJai committed:
feat: add selenium tests for adding data and redirection

useri4l committed:
Merge pull request #56 from AaronJai/sharepage_leaderboard

Sharepage leaderboard Pull Request

AaronJai committed:
feat: update selenium login test to support sign in testing too
- and adjusted naming of tests

AaronJai committed:
fix: add missing import to unittest and update messages model to handle deprecation warning

AaronJai committed:
fix: change email wasn't working
- the original /confirm_email logic only supports tokens generated for new user sign-ups.
- need to create new route for confirming new email addresses

AaronJai committed:
feat: add selenium test to change user details
- change name, email, and password

AaronJai committed:
feat: add selenium test for deleting user

AaronJai committed:
chore: update testing readme and remove unecessary comment in unittest

AaronJai committed:
refactor: convert sign up and log in into reusable helper function

AaronJai committed:
chore: update readme wording and highlight important parts of .env file

useri4l committed:
feat: add custom horizontal legend below charts

- Fixed bug causing chart rendering issues
- Added gradient background to chart section

useri4l committed:
fix: adjust layout to avoid legend being hidden

useri4l committed:
style: apply bold styling to legend text

useri4l committed:
cleanup: delete unnecessary image files

useri4l committed:
Merge branch 'testing' into viewdata_rank

AaronJai committed:
fix: update routes to use main blueprint

useri4l committed:
fix: make tab buttons visible and allow chart reset

Commit:
fix: make tab buttons visible on smaller screen

MeiLin99999 reviewed (approved):
Looking good! Thanks for the tests added to check the security of the website.

AaronJai committed:
style: update axis for leaderboard comparison for consistency and readability
- added y-axis title, increase font size of y-axis scale, increased axis border width

AaronJai committed:
fix: update mobile nav bar to use endpoint splitting and locking logic
- forgot to apply to the header; was from refactoring codebase to support testing and blueprints
- without this the active page won't be highlighted

AaronJai committed:
fix #64: instead of resizing directly, we destroy and re-create the chart each time the window resizes or orientation change

AaronJai committed:
feat: update selenium testing to natively start server by itself instead of requiring it to be open
- better for 'developer experience'
- Matthew also said we'd lose a mark if we don't do this

AaronJai committed:
chore: replace deprecated API usage with new method for SQLAlchemy

AaronJai committed:
feat: update selenium test to use TestingConfig and create database before starting server thread

AaronJai committed:
refactor: replace time.sleep with WebDriverWait and add server polling
- drastically optimises Selenium test runtime (from ~40s to ~17s for all tests up to test_07)
- polling improves robustness across a wide range of devices, including lower-end/slower machines

AaronJai committed:
Merge pull request #61 from AaronJai/viewdata_rank

- Changed the shared section into a new Leaderboard section.
- Users can now click on a name in the leaderboard to compare their carbon data.
- Added a summary and suggestions below each chart to explain the data.
- Added legends to charts to make users easier to understand.
- plus some bug fixes

AaronJai committed:
refactor: modularise routes.py by moving large functions into controllers.py
- helps improve code organisation and maintainability

AaronJai committed:
refactor: update auth page script imports to use flask template

AaronJai committed:
refactor: move scripts for all files into their accompanying static JS file and or Jinja head template

AaronJai committed:
feat: add some selenium tests for facts and dashboard pages

AaronJai committed:
chore: update requirements.txt to include requests

AaronJai committed:
refactor: isolate DB per test and handle deferred JS
- Drop and recreate all tables before each test to prevent data collisions
- Add robust Selenium waits to support deferred JS script loading
- Ensures reliable, isolated, and timing-safe UI tests after JS refactor
- plus some Q.O.L changes for logout button

MeiLin99999 reviewed (approved):
Have checked the files and the selenium testings. Everything in order so far.

AaronJai commented:
> Anything I can help now? @AaronJai

considering i've done almost everything at this point the best thing you can do is reclone the repository and try your best to find any app-breaking bugs.

I'm also going to merge this into main now.
